### ISSUE TITLE
The decision to remove the Gulp 3 typings in favor of Gulp 4 is problematic.

### ISSUE THREAD

**Issue by lddubeau** [DefinitelyTyped/DefinitelyTyped#17183]
The Gulp 3 typings were removed by PR #16121.

It creates two problems.

### `npm install` divergence

As of version 4.x of `@types/gulp` doing `npm install gulp @types/gulp` is a guaranteed recipe for failure because gulp 3 is installed but the types correspond to gulp 4, which restructured the API incompatibly with 3. I mean something as banal as `gulp.task('foo', ['bar'], () => {...})` is not valid in gulp 4.

This is a departure from my experience with DefinitelyTyped so far. By and large, it's been my experience that if `foo` and `@types/foo` both exist, then `npm install foo @types/foo` will produce something that allows my code to compile and run fine.

Gulp 4 does not seem anywhere close to being released:

- There's been no formal alpha, beta or rc release. (By "formal", I mean installable with `npm install gulp@...-alpha`.)

- The work still to be done is substantial. (See http://107.170.66.219/425353. That's their project management board. The issue [https://github.com/gulpjs/gulp/issues/1936] contains some background about that address.)

### Fixing problems in the Gulp 3 typings

The issue above can be worked around with `npm install @types/gulp@3`. However, the Gulp 3 typings have problems. How are we to fix them now?

Pinging @mhegazy who approved the merge and @GiedriusGrabauskas who submitted the PR.

**Comment by leonard-thieu** [comment]
I've put in a [PR](https://github.com/DefinitelyTyped/DefinitelyTyped/pull/17193) for the fix. It looks like this was caused due to a combination of https://github.com/DefinitelyTyped/DefinitelyTyped/commit/a17e430cf3ebcdb0d9cdcb832e418d2bb146c322 (which bumped the major version number without creating a `v3` directory) and the above mentioned PR (which introduced breaking changes from v4).

**Comment by leonard-thieu** [comment]
I've also found out that `npm install @types/gulp@3` isn't good enough. The last release of `@types/gulp` v3 series is 3.8.32 which depends on `@types/orchestrator` and `@types/orchestrator` has introduced changes since then. You'll actually need to install version `4.0.3` to get the latest version of the v3 typings (`npm install @types/gulp@4.0.3 --save-exact`) and drop back down to the v3 series after the PR goes through.

**Comment by DovydasNavickas** [comment]
Not many packages have this kind of situation with absolutely incompatible APIs.
The `@types/gulp` v4 before were incorrect and @GiedriusGrabauskas did a great job writing proper ones.
Is keeping `v3` and `v4` types together as simple as @leonard-thieu did in the PR?
And will it be possible to update them there?

**Comment by leonard-thieu** [comment]
https://github.com/DefinitelyTyped/DefinitelyTyped/pull/16854 is an example of this happening. I added type definitions for `electron-settings` 2.2 after 3.0 already existed.

```
> npm view @types/electron-settings@2

{ name: '@types/electron-settings',
  description: 'TypeScript definitions for electron-settings',
  'dist-tags': 
   { latest: '3.0.0',
     'ts2.0': '3.0.0',
     'ts2.1': '3.0.0',
     'ts2.2': '3.0.0',
     'ts2.3': '3.0.0',
     'ts2.4': '3.0.0',
     'ts2.5': '3.0.0' },
  versions: [ '2.2.0', '3.0.0' ],
  maintainers: [ 'types <ts-npm-types@microsoft.com>' ],
  time: 
   { modified: '2017-06-13T17:15:04.876Z',
     created: '2017-04-21T22:30:31.427Z',
     '3.0.0': '2017-04-21T22:30:31.427Z',
     '2.2.0': '2017-06-01T06:43:16.951Z' },
  repository: 
   { type: 'git',
     url: 'https://www.github.com/DefinitelyTyped/DefinitelyTyped.git' },
  contributors: [ 'Leonard Thieu (https://github.com/leonard-thieu)' ],
  license: 'MIT',
  readmeFilename: '',
  version: '2.2.0',
  main: '',
  scripts: {},
  dependencies: { '@types/node': '*' },
  peerDependencies: {},
  typesPublisherContentHash: '8146f1cab090169a726ebbd0188e7c2bb250dabe8e59d24bbfaab73021de4b8c',
  typeScriptVersion: '2.2',
  dist: 
   { integrity: 'sha512-ra84vlEWUqZ0l4t0cobWTRVo/YWLzgKjlgou99MVL+Q33n7Pf2lDxdHOY7URUHV3rRiXGJ1zqlRWGzbWtvHL6A==',
     shasum: '4a9eb41e2cdef57bccc54414ae1a0c5804cab5d6',
     tarball: 'https://registry.npmjs.org/@types/electron-settings/-/electron-settings-2.2.0.tgz' },
  directories: {} }
```

Although `npm install @types/electron-settings` still installs 3.0 by default even if I have `electron-settings` 2.2 installed, so users will still have to install using `npm install @types/gulp@3`. I don't know enough about npm to know if there's a better way to do this.

**Comment by lddubeau** [comment]
@DovydasNavickas

> Is keeping v3 and v4 types together as simple as @leonard-thieu did in the PR?

I think that's about as good as it gets, given the current constraints under which DefinitelyTyped operates. Having a `v[number]` subdirectory for older versions is a common pattern among the `@types/...` packages.

However, unless there's some feature I don't know, everybody who's using the latest stable release of gulp (until version 4 becomes the latest stable release, some day) will have to add something to their `tsconfig.json` to direct `tsc` to load the definition file in the `v3` subdirectory *instead of* the top level definition file provided by the `@types/gulp` package.

> And will it be possible to update them there?

There should not be any major obstacle to performing updates to the v3 definitions.

**Comment by leonard-thieu** [comment]
`types-publisher` actually produces two separate packages from that structure. v4 won't include the v3 definitions and v3 should look like any other `@types` package.

**Comment by DovydasNavickas** [comment]
As I understand, the whole problem arose because someone added v3 typings as v4.0.3 and now that @GiedriusGrabauskas updated them to proper ones on version v4.0.4, everyone who installed the version `^4.0.3` is out of luck, because it's going to update to the newest one.

As @leonard-thieu poited out, the solution is simple:
```
npm install @types/gulp@3
```

And you do not need to add anything to `tsconfig.json` to make it work, @lddubeau.

**Comment by leonard-thieu** [comment]
```json
'3.8.33': '2017-06-16T16:39:27.892Z'
```
Working v3 typings are now published.
```
npm install @types/gulp@3
```
Don't forget `--save-dev` or `--save` to update your `package.json` as necessary.

**Comment by sandersn** [comment]
Fixed by #17193.
